### ISSUE TITLE
Facility for Grammar Reuse (e.g. Xerces GrammarPool)

### ISSUE THREAD

**Issue by adamretter** [FasterXML/woodstox#20]
@cowtowncoder I am trying to switch eXist-db from Xerces to Woodstox.

Xerces has the facility to setup a shared Grammar Pool, by setting the property `http://apache.org/xml/properties/internal/grammar-pool` and providing an `org.apache.xerces.xni.grammars.XMLGrammarPool` so that parsed Grammars (DTD, XML Schema etc) can be reused without the overhead of re-parsing them. I was wondering if Woodstox (or its use of MSV) had anything similar, or if it was even desirable to have that in Woodstox?

If it is missing but desirable, and you let me know how you envisage it being added, I am happy to do the work and send a PR...

**Comment by cowtowncoder** [comment]
@adamretter Hi there! Yes, Woodstox does actually cache DTDs already. I forget class names, but it should be relatively obvious; but if you can't find it let me know and I'll dig it up.

Other schemas are not auto-resolved at this point, so there is no caching (as there's nothing to really use it; caller has to provide XML Schema / RelaxNG validator). By auto-resolution I just mean that XML Schema mechanism would be detected from the document (I think there was an optional mechanism for doing that, although somewhat frowned upon as there's the view that there isn't really "one schema", but potentially multiple views, for any given document).

Anyway, long story short: for DTDs, caching should exist and work. If not, that could be  a bug.

Having said that, improvements are always welcome, esp. if they could use existing standard API for plugging things in (may or may not be possible). Or otherwise improve things.
